src/xmessage_builder.cpp

```diff
@@ -0,0 +1,48 @@
+#include <string>
+#include <vector>
+
+#include "nlohmann/json.hpp"
+
+#include "xmessage_builder.hpp"
+
+namespace nl = nlohmann;
+
+namespace xeus
+{
+    nl::json create_error_reply(const std::string& evalue,
+                                const std::vector<std::string>& trace_back)
+    {
+        nl::json content;
+        content["status"] = "error";
+        content["ename"] = "Error";
+        content["evalue"] = evalue;
+        content["traceback"] = trace_back;
+        return content;
+    }
+
+    nl::json create_error_reply(const std::string& ename,
+                                const std::string& evalue,
+                                const std::vector<std::string>& trace_back)
+    {
+        const std::vector<std::string>& default_trace_back = nl::json::array();
+        return create_error_reply(ename, evalue, default_trace_back);
```

Comment:
This seems to result in an infinite call stack.

Reply:
Hum...
I thought that was the right way of defining a default value for a const vector?

Reply:
Kinda like what's happening here:

```
    void xkernel_core::send_reply(const std::string& reply_type,
                                  nl::json metadata,
                                  nl::json reply_content,
                                  channel c)
    {
        send_reply(get_parent_id(c),
                   reply_type,
                   get_parent_header(c),
                   std::move(metadata),
                   std::move(reply_content),
                   c);
    }

    void xkernel_core::send_reply(const guid_list& id_list,
                                  const std::string& reply_type,
                                  nl::json parent_header,
                                  nl::json metadata,
                                  nl::json reply_content,
                                  channel c)
    {
        zmq::multipart_t wire_msg;
        xmessage reply(id_list,
                       make_header(reply_type, m_user_name, m_session_id),
                       std::move(parent_header),
                       std::move(metadata),
                       std::move(reply_content),
                       buffer_sequence());
        p_logger->log_sent_message(reply, c == channel::SHELL ? xlogger::shell : xlogger::control);
        std::move(reply).serialize(wire_msg, *p_auth, m_error_handler);
        if (c == channel::SHELL)
        {
            p_server->send_shell(wire_msg);
        }
        else
        {
            p_server->send_control(wire_msg);
        }
    }
```

Reply:
What I mean that this line is calling again `create_error_reply` with three argument, the exact function in which you are already. It's calling itself infinitely.

Reply:
I guess what you wanted to do was to call the `create_error_reply` taking three arguments inside the `create_error_reply` taking two arguments

Reply:
> taking three arguments inside the create_error_reply taking two arguments

Sorry, I don't get it.
Can you explain further?
I thought that was the way to go to overwrite const values on function arguments.

Reply:
Maybe start by writing the tests 😊 it will help you understand it better.

Reply:
@marimeireles you don't want to overwrite the default values, you want to provide a default value when one is missing, eg:

```cpp
nl::json create_error_reply(const std::string& ename,
                                const std::string& evalue,
                                const std::vector<std::string>& trace_back)
{
     // One single implementation to rule them all ... ;)
}

nl::json create_error_reply(const std::string& ename,
                                const std::string& evalue))
{
    std::vector<std::string> default_trace_back = nl::json::array();
    return create_error_reply(ename, evalue, nl::json::array());
}
```